core/src/main/java/org/apache/druid/java/util/common/parsers/JSONFlattenerMaker.java

```diff
@@ -53,16 +53,24 @@
                    .options(EnumSet.of(Option.SUPPRESS_EXCEPTIONS))
                    .build();
 
+  private final boolean keepNullValues;
+
   private final CharsetEncoder enc = StandardCharsets.UTF_8.newEncoder();
 
+  public JSONFlattenerMaker(boolean keepNullValues)
+  {
+    this.keepNullValues = keepNullValues;
+  }
+
   @Override
   public Iterable<String> discoverRootFields(final JsonNode obj)
   {
     return FluentIterable.from(() -> obj.fields())
                          .filter(
                              entry -> {
                                final JsonNode val = entry.getValue();
-                               return !(val.isObject() || val.isNull() || (val.isArray() && !isFlatList(val)));
+                               // If the keepNullValues is set on the JSONParseSpec then null values should not be filtered out
```

Comment:
```suggestion
                               // Root fields are any json elements that ... 
```

I meant something like spelling out how we discover root fields. There are a bunch of nots and ands and ors that I can't come up with a statement that clearly defines what this is trying to do.

This really isn't part of your change, but if you figured out what this is actually doing, you'll help all future devs with a comment